### PR TITLE
Add Constructor Function

### PR DESCRIPTION
**Pull request by dangell7**
There was no easy way to deploy a contract with a signed raw transaction.

How was it fixed?

By adding the deploy_data function. It can be used like so....

`

    greeter_compiled_sol = compile_files(['contracts/Greeter.sol'])
    greeter_contract_interface = greeter_compiled_sol['contracts/Greeter.sol:Greeter']
    greeter_contract = w3.eth.contract(abi=greeter_contract_interface['abi'], bytecode=greeter_contract_interface['bin'])

    contract_data = greeter_contract.deploy_data()

    print(contract_data)

    key = decrypt_account(w3.eth.coinbase, 'Coinbase-Master')

    gas_price = w3.eth.gasPrice
    gas_limit = 7000000

    nonce = w3.eth.getTransactionCount(w3.eth.coinbase)
    
    transaction = {
        'nonce': nonce,
        'gasPrice': gas_price,
        'gas': gas_limit,
        'to': '',
        'value': 1000,
        'data': contract_data,
        'chainId': None
    }
    
    
    signed = w3.eth.account.signTransaction(transaction, key)

    raw_tx_hex = Web3.toHex(signed.rawTransaction)

    print('Raw Tx Hex: %s' % raw_tx_hex)

    tx_hash = w3.eth.sendRawTransaction(raw_tx_hex)
    
    while w3.eth.getTransactionReceipt(tx_hash) == None:
        time.sleep(5)
        print('Waiting on Tranasaction')

    tx_receipt = w3.eth.getTransactionReceipt(tx_hash)

    contract_address = tx_receipt['contractAddress']

    print('Greeter Contract Address: {}'.format(contract_address))

`

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

**Comment by pipermerriam** [comment]
@harpangell7 this feature is captured here in https://github.com/ethereum/web3.py/issues/577 which defines an alternate API for this functionality which is more inline with the existing API for contract functions and events.

Please note that @zheli has claimed that issue, however, there is no open pull request at this time.  I would encourage you to reach out to @zheli prior to investing time in converting this PR to match #577 (assuming you are interested in claiming the bounty).

**Comment by pipermerriam** [comment]
Also, for contributions to be something we can merge, this would need to have tests and documentation written to cover the new functionality.

**Comment by harpangell7** [comment]
Honestly man, I'm not sure I feel comfortable asking for money lol. I just did it because I needed the functionality. I would hate to claim a bounty and not be able to produce. I will keep your tests and documentation in mind though. And I really do appreciate all the work that YOU do!! I'll just use the new API when its released, until then, I have my fork. lol

**Comment by pipermerriam** [comment]
I only mention the bounty to be sure there's as little contention as possible.  You are welcome to contribute the feature of your own altruistic/whatever motivation.  Either way works.

**Comment by harpangell7** [comment]
Well I needed to learn the tests anyways. So I added them and the documentation. But now my "check" with Travis it fails. :( Though, the tests dont assert the deploying to the chain, only that they can sign and return the correct signed transaction details. I couldn't find the tests on the sendRawTransaction().

**Comment by zheli** [comment]
@harpangell7 I have only done some research on current library interface and how it can be done, there is no serious code have been written yet. I would be happy to let you claim this issue (https://github.com/ethereum/web3.py/issues/577) if you want to 👍

**Comment by harpangell7** [comment]
Yeah I'd love to give it a try!! Give me 2 days, and if I cant figure it out, then I'll need some help.

**Comment by zheli** [comment]
@harpangell7 I don't know how to say this, but I actually made some progress on my pull request today... maybe we can work on it together and share the bounty? 😂

**Comment by harpangell7** [comment]
Thats fine, I'm working on the constructor api now. So if you want to tackle the fallback and we can meet in the middle?

**Comment by zheli** [comment]
@harpangell7 I think one of us should branch off from the other's branch to make it easier to track changes:) Maybe you can also tell me briefly how do you want the new constructor to work?

Right now I just need to ask @pipermerriam or @carver what do they mean by deprecating `contract.constructor().deploy` :)

**Comment by harpangell7** [comment]
@zheli go ahead and take this one. I was working on it but got busy and you claimed it first. But to answer your question I think they just mean deprecating contract.deloy() and making the contract.constructor().transact() the new deploy method. I'm curious to see how it turns out though!

**Comment by harpangell7** [comment]
@zheli I made a push to show you how I would do it.

**Comment by zheli** [comment]
@harpangell7 thx man! Regarding to `constructor(*,**)`, I am not sure if they mean to create a new `constructor` method in `Contract` class. As constructor is also a concept in object-oriented programming, in this case it can also mean `__init__()`. At least that's how I understand it.

**Comment by harpangell7** [comment]
Oh interesting, I was thinking they meant the literal constructor, as contract(abi, bytecode) already __init__()'s the class, therefore, calling the constructor on the class made more sense to me. See the language is def lost. lol I also read the fallback as literally contract.fallback(*args, **kwargs).estimateGas(). This would eliminate the need to check for an actually function called fallback, because that function would be called using contract.functions.fallback(*args, **kwargs).estimateGas(). But its really up to them. lol

File: web3/contract.py [hunk 1]
```diff
@@ -268,6 +287,52 @@ def deploy(cls, transaction=None, args=None, kwargs=None):
         txn_hash = cls.web3.eth.sendTransaction(deploy_transaction)
         return txn_hash
 
+    @classmethod
+    @deprecated_for("contract.<constructor>.buildTransaction")
+    def deploy_data(cls, transaction=None, args=None, kwargs=None):
```
Comment (pipermerriam) [inline web3/contract.py]: Can we remove this method since it was never part of the original API.

File: web3/contract.py [hunk 2]
```diff
@@ -1039,6 +1195,93 @@ def call_contract_function(abi,
         return normalized_data
 
 
+# Transact Constructor
+# TODO
```
Comment (pipermerriam) [inline web3/contract.py]: leftover comment?

File: web3/contract.py (hunk 1 above)
Comment (harpangell7) [reply web3/contract.py]: Yes

File: web3/contract.py (hunk 2 above)
Comment (harpangell7) [reply web3/contract.py]: Sorry, I leave my functions as TODO's until finished.

File: tests/core/contracts/test_contract_constructor.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+            '0xf8f68080827c6d94f0109fc8df283027b6285cc889f5aa624eac1f5580b8966060604052604051602080607683395060806040525160008054600160a060020a031916821790555060428060346000396000f3606060405260e060020a600035046334664e3a8114601a575b005b603860005473ffffffffffffffffffffffffffffffffffffffff1681565b6060908152602090f300000000000000000000000016d9983245de15e7a9a73bc586e01ff6e08de73725a0724bfb50b52f80211b4292489822437baf018d5e5ea4d9773ae097095eae50f2a03eb784037880d16d062102567f8b9f25b05e450de5ca5e42e19915505030b247',  # noqa: E501
+            '0xc037d56a1a8910453aa4d28bd193a9e91cfc3bb9640fd6df0562921a81ad158b',
+            51697912782560590269279343002095891073040138366570439304470614233188068643058,
+            28367640749105044447798463648099632309297781255732120427438668075853556003399,
+            37,
+        ),
+    ),
+    ids=['web3js_example', '31byte_r_and_s'],
+)
+
```
[the remark this answers is not in the document]
Comment (harpangell7) [reply tests/core/contracts/test_contract_constructor.py]: Will fix

File: tests/core/contracts/test_contract_constructor.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            '0xf901068080827c6d94f0109fc8df283027b6285cc889f5aa624eac1f5580b8a660606040818152806066833960a09052516080516000918255600155603e908190602890396000f3606060405260e060020a600035046388ec134681146024578063d4c46c7614602c575b005b603460005481565b603460015481565b6060908152602090f300000000000000000000000000000000000000000000000000000000000004d2abcd00000000000000000000000000000000000000000000000000000000000026a081ba76a9445046ac16cc504aa179a1a00f83e140d3507c966997d44691ac61f0a078a2094caf04d34b54a5de4867f3e5c40d53f9550420c39153f0305fb46d8263',  # noqa: E501
+            '0xa51361b96f8169375b310c1ad2236afd9b93bcc414f2627a2dd3bf6d0813b955',
+            58677809990784311110938269820876873533474979477406807277121451199701208293872,
+            54563835237590768768054463437637293096105667696261505012227606369069364511331,
+            38,
+        ),
+    ),
+    ids=['web3js_example', '31byte_r_and_s'],
+)
+
```
[the remark this answers is not in the document]
Comment (harpangell7) [reply tests/core/contracts/test_contract_constructor.py]: Will Fix

File: tests/core/contracts/test_contract_constructor.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+            '0xfc0c69bc23d670970e4d7d0eea80016a17ddb55930959df238d05846a5c62bb6',
+            61711471470314300784969137351093467319738708305914557018954217683388040491481,
+            5563005107683891647322860606868955945559777731600276139365376420219547979743,
+            38,
+        ),
+    ),
+    ids=['web3js_example', '31byte_r_and_s'],
+)
+
+
```
[the remark this answers is not in the document]
Comment (harpangell7) [reply tests/core/contracts/test_contract_constructor.py]: Will Fix

File: tests/core/contracts/test_contract_constructor.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    #assert signed.r == r
+    #assert signed.s == s
+    #assert signed.v == v
+    #raw_tx = Web3.toHex(signed.rawTransaction)
+    #assert raw_tx == expected_raw_tx
+    #assert Web3.toHex(signed.hash) == tx_hash
+
+    #account = acct.privateKeyToAccount(private_key)
+    #assert account.signTransaction(txn) == signed
```
[the remark this answers is not in the document]
Comment (harpangell7) [reply tests/core/contracts/test_contract_constructor.py]: Will Finish

File: docs/contracts.rst [hunk 7]
```diff
@@ -104,6 +104,59 @@ Each Contract Factory exposes the following methods.
...
+
+        >>> transaction = {
+        'to': '',
+        'value': 0,
+        'gas': 2000000,
+        'gasPrice': w3.eth.gasPrice,
+        'nonce': w3.eth.getTransactionCount(w3.eth.coinbase),
+        'chainId': None
+        }
+        >>> contract_data = token_contract.constructor(web3.eth.accounts[1], 12345).buildTransaction(transaction)
```
[the remark this answers is not in the document]
Comment (harpangell7) [reply docs/contracts.rst]: Yes, that makes sense and I will add it now

**Comment by harpangell7** [comment]
Gotcha, yeah the tests were definitely not perfected but I have a good place to start now.

**Comment by harpangell7** [comment]
@pipermerriam I'm running into some trouble with the tests now. When I call contract.constructor().buildTransaction() the result transaction doesn't include the "to" or the "nonce" fields. If I then try to pass this into the signTransaction I get the "TypeError: Not all fields initialized". So then if I try to enter the 'to' field, I have it giving me the 'ValueError: Cannot specify `to` for contract deployment' which makes sense. And If I try to pass the nonce, I receive a "Only the keys 'data/from/gas/gas_price/to/value' are allowed.  Got extra keys: 'nonce'" error. I have made tests to assert that I get an error when trying to pass this information as there shouldn't be a deploy to address and the nonce I'm still trying to figure out. (I think it might just be a test thing, I'm checking in production). The nonce is a new error I haven't seen and I'm not sure how to handle it. Furthermore, the transaction v, s r from the transact() and the v, s, r from the buildTransaction() are different. I can confirm that the data is the same. I will continue testing until its production ready, I just wanted to update you.

**Comment by pipermerriam** [comment]
cc @carver ^  might be something we need to address with how sign transaction validates contract deployment transactions.

Regarding signing the transactions.  As for tests, I don't think you should need to sign the transaction as you should be able to simply submit it over `web3.eth.sendTransaction` without signing it.  Maybe I'm missing something.

**Comment by harpangell7** [comment]
@pipermerriam You're absolutely correct. WOW lol yeah I'll test that @carver when signing the transaction, the nonce error is just a test error, and it will automatically pre populate the "to" as "". So we should be all good. I'll confirm the tests and update the documentation.

**Comment by carver** [comment]
> it will automatically pre populate the "to" as ""

Cool that's roughly the approach I was going to suggest @harpangell7 . My instinct is that `to` should be `None`, and if that doesn't work in `signTransaction` then I'd like to update that function.

**Comment by harpangell7** [comment]
@pipermerriam @carver So buildTransaction() adds the default chainId and this is giving me a "eth_tester.exceptions.ValidationError: Only the keys 'data/from/gas/gas_price/nonce/to/value' are allowed.  Got extra keys: 'chainId'". when I try to run web3.eth.sendTransaction. I've confirmed that this is only a testing error, as I can use buildTransaction({'chainId': 1}) just fine in my program.

**Comment by harpangell7** [comment]
@carver So I was incorrect in stating that it will auto 'to' to "". The reason it was doing that was because I hacked in built_transaction['to'] = '' just before returning to the user. Will there ever be an instance when they can deploy to a specific address? Not to sound stupid, but I want to make sure. With the current hack though, when you sign the transaction, it changes the "" to None.

**Comment by carver** [comment]
> buildTransaction() adds the default chainId and this is giving me a "eth_tester.exceptions.ValidationError: Only the keys 'data/from/gas/gas_price/nonce/to/value' are allowed. Got extra keys: 'chainId'"

Can you open an issue on `eth-tester` about this (and link back here)?

> Will there ever be an instance when they can deploy to a specific address?

Nope, the defined way to deploy a contract is to leave the `to` field empty. So building a transaction from a constructor must set `to` to empty.

**Comment by harpangell7** [comment]
@carver Purrfect!

**Comment by dangell7** [comment]
@carver I’m just waiting on the eth-tester issue before I submit again...

**Comment by nueverest** [comment]
Would love to see this working.

**Comment by carver** [comment]
@nueverest Please subscribe to #666 which will be updated when the problem is fixed (even if this particular PR doesn't get merged)

**Comment by carver** [comment]
Resolved in #693

File: web3/contract.py [hunk 8]
```diff
@@ -721,6 +746,99 @@ def __call__(self, *args, **kwargs):
...
+            built_transaction = dict(**transaction)
+
+        built_transaction = build_transaction_for_constructor(self.contract_abi,
+                                                 self.contract_bytecode,
+                                                 self.web3,
+                                                 built_transaction,
+                                                 *self.args,
+                                                 **self.kwargs)
+
+        built_transaction['to'] = ''
```
Comment (dangell7) [inline web3/contract.py]: Here was my suggestion to fix that bug. https://github.com/ethereum/web3.py/issues/739

**Pull request closed without merging**